tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc

```diff
@@ -254,68 +227,70 @@ bool CouldNeedLibdevice(const llvm::Module& module) {
   return false;
 }
 
-// Links libdevice into the given module if the module needs libdevice.
-Status LinkLibdeviceIfNecessary(llvm::Module* module,
-                                std::pair<int, int> compute_capability,
-                                const string& libdevice_dir_path) {
-  if (!CouldNeedLibdevice(*module)) {
-    return Status::OK();
-  }
-
-  // CUDA 9+ uses a single libdevice file for all devices, and we don't support
-  // older CUDAs.
-  string libdevice_path =
-      tensorflow::io::JoinPath(libdevice_dir_path, "libdevice.10.bc");
-  if (!tensorflow::Env::Default()->FileExists(libdevice_path).ok()) {
-    LOG(WARNING)
-        << "libdevice is required by this HLO module but was not found at "
-        << libdevice_path;
-    return xla::InternalError("libdevice not found at %s", libdevice_path);
-  }
+// Links the module with a vector of path to bitcode modules
+// The paths are guaranteed to exist.
+Status LinkWithBitcodeVector(llvm::Module* module,
+                             const std::vector<string>& bitcode_path_vector) {
+  llvm::Linker linker(*module);
 
-  VLOG(1) << "Linking with libdevice from: " << libdevice_path;
-  std::unique_ptr<llvm::Module> libdevice_module =
-      LoadIRModule(libdevice_path, &module->getContext());
+  for (auto& bitcode_path : bitcode_path_vector) {
+    if (!tensorflow::Env::Default()->FileExists(bitcode_path).ok()) {
+      LOG(WARNING) << "bitcode module is required by this HLO module but was "
+                      "not found at "
+                   << bitcode_path;
+      return xla::InternalError("bitcode module not found at %s", bitcode_path);
+    }
 
-  llvm::Linker linker(*module);
-  if (linker.linkInModule(
-          std::move(libdevice_module), llvm::Linker::Flags::LinkOnlyNeeded,
-          [](Module& M, const StringSet<>& GVS) {
-            internalizeModule(M, [&GVS](const GlobalValue& GV) {
-              return !GV.hasName() || (GVS.count(GV.getName()) == 0);
-            });
-          })) {
-    return xla::InternalError("Error linking libdevice from %s",
-                              libdevice_path);
+    std::unique_ptr<llvm::Module> bitcode_module =
+        LoadIRModule(bitcode_path, &module->getContext());
+    if (linker.linkInModule(
+            std::move(bitcode_module), llvm::Linker::Flags::LinkOnlyNeeded,
+            [](Module& M, const StringSet<>& GVS) {
+              internalizeModule(M, [&M, &GVS](const GlobalValue& GV) {
+                return !GV.hasName() || (GVS.count(GV.getName()) == 0);
+              });
+            })) {
+      return xla::InternalError("Error linking bitcode module from %s",
+                                bitcode_path);
+    }
   }
   return Status::OK();
 }
 
-StatusOr<string> CompileModuleToPtx(llvm::Module* module,
-                                    std::pair<int, int> compute_capability,
+StatusOr<std::unique_ptr<llvm::TargetMachine>>
+ConstructLLVMTargetMachineForModule(llvm::Module* module,
```

Comment:
The function name ConstructLLVMTargetMachineForModule seems misleading and sound similar to GetTargetMachine. The most important thing this function does is running assembling and running compiler passes on the module which is an output param. This needs to be reflected in the name. The target_machine is more of a by-product. I would be cleaner to pass target_machine in as param.

Reply:
How do you think if I rename this function from `ConstructLLVMTargetMachine` to something like `LinkAndOptimizeModule`?